### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by haskellcamargo**
This pull-request is mostly refactoring. It does:

- Add linter for PSR-2 with `phpcs`

Unfortunately, I wasn't able to configure on Travis, as long as Pyrus and Pear are no longer supported by Travis

- Remove unused `.qk` files
- Fix all PHP code to be compliant with PSR-2
- Remove `pub` keyword
- Remove `rec` keyword
- Remove pass-by-reference and references
- Remove nullable member access with `?.`
- Remove clone operator `^^`
- Create a `FnSignatureStmt` for general use
- Remove dead code
- Add test for type signature
- Remove `T_DYN` and `T_RESOURCE` (both unused)
- Remove symbol table for lexer. It doesn't need that

### CODE REVIEW

File: src/ast/Node.php [hunk 1]
```diff
@@ -70,12 +70,7 @@ private function bindDecl($named_node, $type, $kind)
             throw new ScopeError(Localization::message('SCO130', [$named_node->name, $type]));
         }
 
-        // When it is a function and it is marked as public
-        if (Kind::K_FUNCTION === $kind && $named_node->is_pub) {
```
Comment (luizperes) [inline src/ast/Node.php:74]: You will add a keyword to replace the `pub` keyword, right? In this case, do you think that it is worth to delete it completely?

File: src/ast/expr/LambdaExpr.php [hunk 2]
```diff
@@ -52,18 +52,21 @@ public function format(Parser $parser)
                 break;
             case 1:
                 if ($this->has_brackets) {
-                    $source .= '['
-                             . ($this->parameters[0]->is_reference ? '*' : '')
```
Comment (luizperes) [inline src/ast/expr/LambdaExpr.php:56]: (y)

File: src/ast/expr/LambdaExpr.php [hunk 3]
```diff
@@ -52,18 +52,21 @@ public function format(Parser $parser)
...
                     $source .= $this->parameters[0]->name;
                 }
                 break;
             default:
                 $source .= '[';
                 $source .= implode(', ', array_map(function ($param) {
-                    return ($param->is_reference ? '*' : '') . $param->name;
+                    $parameter = $param->name;
+
+                    if (isset($param->type)) {
```
Comment (luizperes) [inline src/ast/expr/LambdaExpr.php]: using `isset` here doesn't go against our current member checking? `NULL !== $param->type`
Or is that another reason that I did not see? :) just wondering ^^

File: src/ast/stmt/ExprStmt.php [hunk 4]
```diff
@@ -66,6 +66,8 @@ public function runTypeChecker()
             $type_list[] = $expr->getType();
         }
 
-        var_dump(array_map(function ($type) { return (string) $type; }, $type_list));
+        var_dump(array_map(function ($type) {
```
Comment (luizperes) [inline src/ast/stmt/ExprStmt.php:69]: this `var_dump` is supposed to be there? or the `ExprStmt` is not done yet?

File: src/ast/stmt/FnSignatureStmt.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        $source = '';
+        if ($this->native) {
+            $source .= 'native fn ';
+        }
+        $source .= $this->name;
+        $source .= '(';
+        $source .= implode(', ', array_map(function ($param) {
+            $parameter = $param->name;
+
+            if (isset($param->type)) {
```
Comment (luizperes) [inline src/ast/stmt/FnSignatureStmt.php]: I think that it is the same thing here for the `isset`

File: src/ast/stmt/ForeachStmt.php [hunk 6]
```diff
@@ -114,7 +108,7 @@ public function runTypeChecker()
         $generator_type = $this->generator->getType();
 
         // When the element has no subtype to be an iterable
-        if (null === $generator_type->subtype)  {
+        if (null === $generator_type->subtype) {
```
Comment (luizperes) [inline src/ast/stmt/ForeachStmt.php:111]: here we have the other pattern

File: src/scope/Meta.php [hunk 7]
```diff
@@ -23,9 +23,9 @@
 
 class Meta
 {
-    const M_REF_COUNT = "refcount";
-    const M_TYPE      = "type";
-    const M_LABEL     = "label"; // implicit meta labels
+    const M_REF_COUNT = 'refcount';
```
Comment (luizperes) [inline src/scope/Meta.php:26]: I always forget >.<

**Review by luizperes — COMMENTED** [review]
I believe that we have some minor fixes (that I have already pointed out), but other than that, it is pretty good! :D It is good to go!

**Review by luizperes — APPROVED** [review]

File: src/ast/Node.php (hunk 1 above)
Comment (haskellcamargo) [reply src/ast/Node.php:74]: Yes, but adding keywords is quite easy, and cleaning dead code may be better than letting it skip.

File: src/ast/expr/LambdaExpr.php (hunk 3 above)
Comment (haskellcamargo) [reply src/ast/expr/LambdaExpr.php]: You are right!

File: src/ast/stmt/ExprStmt.php (hunk 4 above)
Comment (haskellcamargo) [reply src/ast/stmt/ExprStmt.php:69]: Yes, because I'm using it to debug type inference on REPL.

File: src/ast/stmt/FnSignatureStmt.php (hunk 5 above)
Comment (haskellcamargo) [reply src/ast/stmt/FnSignatureStmt.php]: Right again.

**Pull request merged**